### PR TITLE
ci: setup benchmark test environment on equinix metal

### PR DESCRIPTION
**Pull request by yangchiu**
ci: setup benchmark test environment on equinix metal

For https://github.com/longhorn/longhorn/issues/3885

Signed-off-by: Yang Chiu <yang.chiu@suse.com>

### CODE REVIEW

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Some comments needed to resolve.

File: benchmark_test/Jenkinsfile [hunk 1]
```diff
@@ -41,7 +45,13 @@ node {
                                    --env TF_VAR_lh_aws_instance_name_worker="${JOB_BASE_NAME}-wrk" \
                                    --env TF_VAR_lh_aws_instance_type_controlplane=${CONTROLPLANE_INSTANCE_TYPE} \
                                    --env TF_VAR_lh_aws_instance_type_worker=${WORKER_INSTANCE_TYPE}\
+                                   --env TF_VAR_instance_name_controlplane="${JOB_BASE_NAME}-ctrl" \
+                                   --env TF_VAR_instance_name_worker="${JOB_BASE_NAME}-wrk" \
+                                   --env TF_VAR_instance_type_controlplane=${CONTROLPLANE_INSTANCE_TYPE} \
+                                   --env TF_VAR_instance_type_worker=${WORKER_INSTANCE_TYPE}\
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply benchmark_test/Jenkinsfile:51]: benchmark test pipeline shares the same terraform scripts with all daily regression test pipelines.To remove `TF_VAR_lh_aws_instance_xxx`, need to modify the jenkins job definitions of all regression test pipelines too.

I would suggest it to be done in another separate task.

File: test_framework/terraform/equinix/ubuntu/user-data-scripts/provision_k3s_agent.sh.tpl [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+set -e
+set -x
+
+apt-get update
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply test_framework/terraform/equinix/ubuntu/user-data-scripts/provision_k3s_agent.sh.tpl:6]: No, the supported systems are listed here: https://deploy.equinix.com/developers/docs/metal/operating-systems/supported/

File: test_framework/terraform/aws/sles/user-data-scripts/provision_k3s_agent.sh.tpl [hunk 3]
```diff
@@ -6,6 +6,20 @@ sudo zypper install -y open-iscsi nfs-client
...
+  mount /dev/nvme1n1 /mnt/sda1
+
+  mkdir /mnt/sda1/local
+  mkdir /opt/local-path-provisioner
+  mount --bind /mnt/sda1/local /opt/local-path-provisioner
+
+  mkdir /mnt/sda1/longhorn
+  mkdir /var/lib/longhorn
+  mount --bind /mnt/sda1/longhorn /var/lib/longhorn
+fi
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply test_framework/terraform/aws/sles/user-data-scripts/provision_k3s_agent.sh.tpl]: refined!

File: test_framework/terraform/equinix/ubuntu/user-data-scripts/provision_k3s_agent.sh.tpl [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+apt-get install -y nfs-common
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply test_framework/terraform/equinix/ubuntu/user-data-scripts/provision_k3s_agent.sh.tpl:7]: No, it's already been installed:
```
# apt list --installed | grep iscsi
libopeniscsiusr/jammy,now 2.1.5-1ubuntu1 amd64 [installed,automatic]
open-iscsi/jammy,now 2.1.5-1ubuntu1 amd64 [installed,automatic]
```

File: test_framework/terraform/equinix/ubuntu/main.tf [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+resource "equinix_metal_device" "control_plane" {
+
+  depends_on = [
+    equinix_metal_ssh_key.ssh_key
+  ]
+
+  hostname         = "${var.instance_name_controlplane}-${random_string.random_suffix.id}"
+  plan             = var.instance_type_controlplane
+  metro            = var.metro
+  operating_system = "ubuntu_${replace(var.os_distro_version, ".", "_")}"
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply test_framework/terraform/equinix/ubuntu/main.tf:46]: No, the supported systems are listed here: https://deploy.equinix.com/developers/docs/metal/operating-systems/supported/

File: benchmark_test/scripts/benchmark-test-setup.sh [hunk 6]
```diff
@@ -8,19 +8,10 @@ TMPDIR="/tmp/longhorn"
...
-    if [[ ${TF_VAR_k8s_distro_name} == [rR][kK][eE] ]]; then
-      export KUBECONFIG="${BASEDIR}/kube_config_rke.yml"
-    elif [[ ${TF_VAR_k8s_distro_name} == [rR][kK][eE]2 ]]; then
-      export KUBECONFIG="${BASEDIR}/terraform/aws/${DISTRO}/rke2.yaml"
-    else
-      export KUBECONFIG="${BASEDIR}/terraform/aws/${DISTRO}/k3s.yaml"
-    fi
-  elif [[ ${ARCH} == "arm64"  ]]; then
-    export KUBECONFIG="${BASEDIR}/terraform/aws/${DISTRO}/k3s.yaml"
+  if [[ "${TF_VAR_k8s_distro_name}" == "rke2" ]]; then
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply benchmark_test/scripts/benchmark-test-setup.sh:11]: If rke1 will not be used anymore, probably we can create a task to remove all rke1 related code.

File: test_framework/terraform/equinix/ubuntu/variables.tf [hunk 7]
```diff
@@ -0,0 +1,78 @@
+variable "k8s_distro_name" {
+  type        = string
+  default     = "rke2"
+  description = "kubernetes distro version to install [rke2, k3s]  (default: rke2)"
+}
+
+variable "k8s_distro_version" {
+  type        = string
+  default     = "v1.25.3+rke2r1"
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply test_framework/terraform/equinix/ubuntu/variables.tf]: I modified it to v1.24.7+rke2r1 first to be consistent with https://github.com/longhorn/infra/pull/108

File: benchmark_test/scripts/benchmark-test-setup.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (innobead) [reply benchmark_test/scripts/benchmark-test-setup.sh:11]: Sounds good to me, but not a priority for now.

**Review by innobead — APPROVED** [review]

**Pull request merged**